tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.Reporter.cs

```diff
@@ -106,14 +106,23 @@ internal void TryReport(IResult result, bool blocked, int? status = null)
 
             LogMatchesIfDebugEnabled(result.Data, blocked);
 
-            _localRootSpan.SetTag(Tags.AppSecJson, "{\"triggers\":" + result.Data + "}");
+            var traceContext = _localRootSpan.Context.TraceContext;
+            if (traceContext.WafSecurityEvents == null)
+            {
+                traceContext.WafSecurityEvents = result.Data;
+            }
+            else if (result.Data != null)
+            {
+                traceContext.WafSecurityEvents.AddRange(result.Data);
```

Comment:
Because we don't copy the list, this is modifying the data that's held in `IResult`.

1. Is that ok _now_?
2. Is that ok in the future? hopefully nothing else enumerates it while this is running 😅

Reply:
It's a good point, it shouldn't happen now, but could happen in the future. I changed the type exposed to `IReadOnlyCollection<T>` to hopefully help future devs understand they shouldn't change it.

It's also possible the `WafSecurityEvents` could be access from two different threads, so changed that to `ConcurrentBag<T>`

See 937873794

Reply:
😱 Not `ConcurrentBag`, @kevingosse will be summoned! 😅 https://minidump.net/performance-best-practices-in-c-b85a47bdd93a/ We should use `ConcurrentQueue` instead (if it does, in fact, need to be thread-safe)

![image](https://github.com/DataDog/dd-trace-dotnet/assets/18755388/507cc18a-b364-46c9-875f-8a1c509056d6)

Also, I don't think we should _always_ allocate the `WafSecurityEvents` collection in the `TraceContext` 🤔 we could allocate it on demand no?

Reply:
Always allocating `WafSecurityEvents` convenient and safe, but you're it's probably overhead we don't want. 

Given the safety checks necessary to safely initialize a field, I feel it's probably easier to fall back to just using a lock. Given contention is rare the overhead should be low. To try and avoid overhead when ASM is disabled, See 70eab30a9 for more details.